cmd/rpcdaemon/commands/get_receipts.go

```diff
@@ -105,34 +105,142 @@ func newFilter(addresses []common.Address, topics [][]common.Hash) *Filter {
 
 // GetLogs returns logs matching the given argument that are stored within the state.
 func (api *APIImpl) GetLogs(ctx context.Context, crit filters.FilterCriteria) ([]*types.Log, error) {
-	var filter *Filter
+	var begin, end uint32
+	var logs []*types.Log //nolint:prealloc
+
+	tx, beginErr := api.dbReader.Begin(ctx)
+	if beginErr != nil {
+		return returnLogs(logs), beginErr
+	}
+	defer tx.Rollback()
+
 	if crit.BlockHash != nil {
-		// Block filter requested, construct a single-shot filter
-		filter = NewBlockFilter(*crit.BlockHash, crit.Addresses, crit.Topics)
+		number := rawdb.ReadHeaderNumber(tx, *crit.BlockHash)
+		if number == nil {
+			return nil, fmt.Errorf("block not found: %x", *crit.BlockHash)
+		}
+		begin = uint32(*number)
+		end = uint32(*number)
 	} else {
 		// Convert the RPC block numbers into internal representations
-		latest, err := getLatestBlockNumber(api.dbReader)
+		latest, err := getLatestBlockNumber(tx)
 		if err != nil {
 			return nil, err
 		}
 
-		begin := int64(latest)
+		begin = uint32(latest)
 		if crit.FromBlock != nil {
-			begin = crit.FromBlock.Int64()
+			begin = uint32(crit.FromBlock.Uint64())
 		}
-		end := int64(latest)
+		end = uint32(latest)
 		if crit.ToBlock != nil {
-			end = crit.ToBlock.Int64()
+			end = uint32(crit.ToBlock.Uint64())
 		}
-
-		filter = NewRangeFilter(begin, end, crit.Addresses, crit.Topics)
 	}
-	// Run the filter and return all the logs
-	logs, err := filter.Logs(ctx, api)
+
+	blockNumbers := gocroaring.New()
+	blockNumbers.AddRange(uint64(begin), uint64(end+1)) // [min,max)
+
+	topicsBitmap, err := getTopicsBitmap(tx.(ethdb.HasTx).Tx().Cursor(dbutils.LogTopicIndex), crit.Topics)
 	if err != nil {
 		return nil, err
 	}
-	return returnLogs(logs), err
+	if topicsBitmap != nil {
+		if blockNumbers == nil {
+			blockNumbers = topicsBitmap
+		} else {
+			blockNumbers.And(topicsBitmap)
+		}
+	}
+
+	logAddrIndex := tx.(ethdb.HasTx).Tx().Cursor(dbutils.LogAddressIndex)
+	var addrBitmap *gocroaring.Bitmap
+	for _, addr := range crit.Addresses {
+		m, err := bitmapdb.Get(logAddrIndex, addr[:])
+		if err != nil {
+			return nil, err
+		}
+		if addrBitmap == nil {
+			addrBitmap = m
+		} else {
+			addrBitmap = gocroaring.Or(addrBitmap, m)
```

Comment:
Why do you use `Or` here and `FastOr` in another place?

Reply:
Or - accept 2 bitmaps, FastOr - many, for 2 it works same as Or. so, no real reason here.